### ISSUE TITLE
[BUG] Shop Receipt Missing Core Shipping Details

### ISSUE THREAD

**Issue by chris-rogala** [etsy/open-api#211]
**Describe the bug**
Two core properties, `shipping_details` and `shipped_date`, in the receipt is missing in v3. 

**To Reproduce**

1) Send the below to v2:

`curl --location --request GET 'https://openapi.etsy.com/v2/receipts/{receiptId}?oauth_consumer_key={key}&oauth_token={token}&oauth_signature_method=HMAC-SHA1&oauth_timestamp=1630355597&oauth_nonce=0thnU4dBTfE&oauth_version=1.0&oauth_signature={signature}'`

2) Send the below to v3:

```
curl --location --request GET 'https://openapi.etsy.com/v3/application/shops/{shopId}/receipts/{receiptId}' \
--header 'x-api-key: {apiKey}' \
--header 'authorization: Bearer {token}'

```
3) Compare the responses. V3 is missing `shipping_details` and `shipped_date` which provide the ship by date and the estimated delivery date.

**Expected behavior**
Have properties in the receipt model that provides the ship-by-date and the estimated delivery date. This is core to our process of managing orders within our system and we cannot move forward with v3 without these properties.

**Screenshots**
![2021-08-30_17-23-59](https://user-images.githubusercontent.com/44238629/131408718-128450c3-a093-41fb-976f-aa9c79bf103c.jpg)

**Additional context**
### **Edit to add context**
This might be considered an enhancement for Etsy, but right now, it will likely cause a bug in our system. We will be pulling the shipping profiles and listings, and we could save that static information when we get the receipt, but the shipping profile could be modified before or after the order is placed, so the processing times could change. Not having these properties static in the receipt could result in an order being shipped late because it was not in our queue correctly.

Also, this brings up another bug from v2. When the ship-by-date is modified, the `update_timestamp` did not update, so when we pulled receipts based on `min_last_modified`, the receipt did not get returned. We had several orders shipped late because of this bug in v2. This also points out another flaw with referencing the shipping profile since when the ship-by-date is updated, the processing window in the profile becomes irrelevant.

**Comment by chris-rogala** [comment]
It appears the bug from v2 is still there. Today, we had to push out the ship by date on order 2159440229 because we did not receive the image in time to create the artwork. We updated the ship by date from today and moved it to 9/20. However, the `update_timestamp` did not change. There isn't a way to update the ship by date in the API, so we don't have a way to make note of the change within our system. Currently, we poll the active orders and determine if the `shipped_date` has been updated. 

Please feel free to look at our shop, and you will see that this will become extremely problematic in a few months. 

Below is a screenshot from Postman. 
 
![2021-08-30_21-21-22](https://user-images.githubusercontent.com/44238629/131426206-d847d6e2-e3e8-4c19-8273-69e3a9ef5abc.jpg)

**Comment by glukoid** [comment]
guys, could you please help me to understand why I could not get receipt by V3 API at all ?

I am able to receive receipts by API V2, but when I am trying to get receipts by call to API V3, the status code is Forbidden
First of all I called to ExchangeOfOAuthTokens from V2 to V3.  ("v3/public/oauth/token")
from https://developers.etsy.com/documentation/essentials/authentication/#exchange-oauth-10-token-for-oauth-20-token

I got successful response (NOTE: V2 legacy token has scopes of transactions_r and transactions_w

Request to receive receipt by API V3:
Resource: $"v3/application/shops/{shopName}/receipts/{orderId}"
Header x-api-key - ApplicationKey
Authorization Bearer + authToken.AccessToken
8) result: {"error":"insufficient_scope","error_description":"Insufficient scope"}

thx

**Comment by chris-rogala** [comment]
@glukoid  Make sure you are also requesting `email_r` in your scope for your token. The receipts require both `transactions_r` and `email_r`. 

![2021-08-31_15-13-05](https://user-images.githubusercontent.com/44238629/131562257-2862f7cd-27d6-411c-8733-e80e3cc74889.jpg)

**Comment by glukoid** [comment]
@chris-rogala 👍 

it helped (I just did not thought the whole receipt will not be returned because of not providing email_r scope)

thank you

**Comment by etsyachristensen** [comment]
Thanks @chris-rogala for jumping in there and helping @glukoid. We added the `email_r` scope to the endpoint because we were asked to include the email addresses for the buyer and seller as part of the receipt. This meant we had to make sure that apps had permission from the shop owner to see that information.

@chris-rogala I will pass on the bug information to the V2 support team if you haven't already. For V3, I am currently working on adding the shipping method selected to the receipt and then will look at adding those additional shipping details as well.

**Comment by chris-rogala** [comment]
@etsyachristensen can you make sure the last modified date is changed when the ship by date is adjusted? This is vital for our application to manage our queue during the holiday season. Our queue can reach over 1000 open orders and this is vital to ensure our reports are accurate and orders are not missed. Right now, we have to pull all open receipts and then compare the ship by date with what we have. This isn’t a big deal with a queue that is under 100 but becomes a bit of an issue when the queue exceeds 1000. Thanks!

**Comment by Killerkiss** [comment]
@etsyachristensen Hi.
I am currently working on integrating your API for my etsy stores. Having almost finished everything, I realized that the receipt does not display information about the delivery method and the timing of shipment and approximate delivery. Please add this information as it is critical for any store.
Please let us know as soon as you add this functionality.
Thanks.

**Comment by chris-rogala** [comment]
@Killerkiss As long as you are not updating your shipping profile often or adjusting the ship by date, you can get close to the ship by date through the shipping profile. You have to get the item, then get the shipping profile. Unfortunately, that information is not static, so if the shipping profile changes before you pull the receipt, then the ship by date will not be accurate. We do adjust ship by dates because we rely on customers sending us photos, and if they don’t, we push their ship by date. We also adjust our shipping profiles to ensure customers have accurate lead times. For us, the shipping profiles are adjusted more frequently in October and November. For now, I am pull that information from v2 and I have stopped our project to transition off of v2 to focus on new functionality needed got this fall. Hopefully this helps.

**Comment by glukoid** [comment]
since you guys collecting all other issues I want to add my 2 cents here:
1) new API does not return the date when an order should be shipped.
2) transaction does not include SKU property. it could be retrieved by additional call, but just want to be sure this was the plan.
3) cancelled order. API V2 returned is_dead  = true for cancelled order, but API V3 returned error message "Could not find a Shop Receipt with receipt_id". Is it the new way of API or a bug ?
4) API V3 does not return does country name. instead it return country_iso. Does that mean country name should retrieved by call to another api ? or we suppose to retrieved it from formatted_address ? BTW what is the reason behind this field ?

I tried to get answers by opening request for development team but they did not reply.

thx

**Comment by Killerkiss** [comment]
@glukoid about your questions on "formatted_address".
I think they use it for quick search field on Etsy shop manager. One field text search instead first_name+second_name+first_line+second_line+country+postcode and etc. I also used this field for search in my own crm system. Hopefully this helps.

**Comment by Killerkiss** [comment]
@chris-rogala thank you for reply. I will use it as temporary method, but i need selected shipping method by current receipt. I really hope that this data will be added soon, since this is literally the only thing that now stops me from starting to use my crm system :)

**Comment by chris-rogala** [comment]
@glukoid I'm not sure if this helps, but hopefully it does. 

2) I think the idea here is to call the receipts endpoint and then do a batch call on the listing endpoint

STEP ONE: GET RECEIPTS

```
curl --location --request GET 'https://openapi.etsy.com/v3/application/shops/{{shop_id}}/receipts?was_shipped=false' \
--header 'x-api-key: {{api_key}}' \
--header 'Authorization: Bearer {{access_token}}'
```

STEP TWO: GET LISTINGS

```
curl --location --request GET 'https://openapi.etsy.com/v3/application/listings/batch?listing_ids={{listing_id_one}},{{listing_id_two}}' \
--header 'x-api-key: {{api_key}}' \
--header 'Authorization: Bearer {{access_token}}'
```

I do agree that adding the SKU makes a lot of sense for this endpoint since that prop is core for sellers.

3) In my opinion, this does make sense to not find the receipts. However, returning a **404** doesn't make sense it is the same status code as an invalid receipt. I think a more appropriate status code would be a **410**. With this status code, our code could recognize that the order did exist but is not longer available. Another options is to move the resource to an archive location and return a **308** with the **Location** header having the archived location.  This way a seller can follow-up with a buyer if the order was accidently cancelled. 

4) They are moving to the Country ISO which is actually a lot better since it uses a standard. Here is where they mentioned the deprecation of the Country API [https://developers.etsy.com/documentation/migration/retired-resources](https://developers.etsy.com/documentation/migration/retired-resources). I populated our own API with the ISO table by referencing this [https://en.wikipedia.org/wiki/List_of_ISO_3166_country_codes](https://en.wikipedia.org/wiki/List_of_ISO_3166_country_codes). Based on the data, it appears they are referencing the alpha2 code, but I wish they would be more clear of which format they are going to be using.

Happy coding!

**Comment by chris-rogala** [comment]
> @chris-rogala thank you for reply. I will use it as temporary method, but i need selected shipping method by current receipt. I really hope that this data will be added soon, since this is literally the only thing that now stops me from starting to use my crm system :)

It's a terrible workaround and you have to make assumptions on the shipping method based on the price. This is why I stopped our migration to v3 as this is core for us managing our queue. If they deprecate v2 during the holiday season, we will likely have to turn off our shop and literally stop taking orders since we would have no way to manage the queue since our queue exceeds 1000 orders in the holiday season.

**Comment by glukoid** [comment]
@chris-rogala
2) I implemented something like that. I got it by listing_id + product_id call to listing api
3) agree if this was the plan of change. but it could be just a bug. e.g. API V2 returned is_dead = true only for implicit call to the receipt of cause it wasn`t very convenient.
4) for now as an ugly workaround I got country name from formatted_address. of cause it is not suppose to work this way. 

but of cause the most urgent issue is the shipping date property for V3 orders :(

thx

**Comment by chris-rogala** [comment]
@glukoid the country code is in the receipt. It’s the `country_iso` property.

![5368D2A5-8545-4B3D-82F2-F37C067D01FD](https://user-images.githubusercontent.com/44238629/133000313-7eecd38d-960d-485f-80de-bf1fe43695cd.jpeg)

**Comment by glukoid** [comment]
@chris-rogala 
yes I know country_iso in the receipt. but I need to receive country name.
so my choice was either to convert coutnry_iso to country name by call to some API or my local DB (assuming I will have ico table) or to get it from formatted_address. I know it is workaround but any way I could not use API V3 without shipping date for an order.

thx

**Comment by timalin** [comment]
> @chris-rogala
> 2) I implemented something like that. I got it by listing_id + product_id call to listing api
> 3) agree if this was the plan of change. but it could be just a bug. e.g. API V2 returned is_dead = true only for implicit call to the receipt of cause it wasn`t very convenient.
> 4) for now as an ugly workaround I got country name from formatted_address. of cause it is not suppose to work this way.
> 
> but of cause the most urgent issue is the shipping date property for V3 orders :(
> 
> thx

for 2), as the product_id is unique, I'm considering associating them with sku locally. there's actually a running issue in v2 when receipts with transactions don't return sku field(return "None"). happens periodically.

**Comment by etsyachristensen** [comment]
@glukoid and @timalin.
1. The country_iso that is returned is the 3 character country code and is internationally standard. You should be able to translate the country name from the code. See more information here: https://www.iso.org/iso-3166-country-codes.html. We have no plans at this time to provide the country name because all of the major shipping companies rely on the ISO code.

2. We are currently working on providing the shipping method and shipping dates as part of the transactions that will be listed in the receipts. We are in the testing phase of this so hopefully this will go into production soon.

3. We have another github issue opened regarding canceled orders. We are still evaluating if/when we will put this into V3. I'd like to see it added as well, but exactly HOW we add it is the question.

4. I will look at options for adding the sku to the receipt as well.

**Comment by stickfigure** [comment]
@etsyachristensen can you confirm (because I couldn't find this in the docs) that it's the 3 letter code code, not the 2 letter code, that Etsy uses in all cases for country_iso?

**Comment by chris-rogala** [comment]
@stickfigure Our orders are coming through as the 2 letter code:

![image](https://user-images.githubusercontent.com/44238629/133677455-cfd5c4bb-6b4a-4cdb-9acf-8d0e26b45740.png)

**Comment by etsyachristensen** [comment]
My apologies @stickfigure and @chris-rogala . I misspoke. I meant 2 letters not 3.

**Comment by etsyachristensen** [comment]
Ok this thread is definitely getting off topic. I'm hoping to move some of this to some new threads next week to continue discussion in siloed topics.

To address the original post, we have now added the following fields to the transaction associations which are on the receipts:
"shipping_method"  - This will be "Standard Shipping" unless a shipping upgrade has been selected by the buyer.
"expected_ship_date" - This will be the date that the item is expected to be shipped out by based on the listing data. Returns null if the item is already shipped.
"actual_ship_date" - This is the date that the item was actually shipped.

I know that there is more data being requested for the receipt endpoints. We are working on those and have put those in separate tickets to be worked on. Hopefully this at least helps a bit.

**Comment by chris-rogala** [comment]
@etsyachristensen what happens if a ship by date is changed on an order? Will these fields change? While I understand the expected ship by date is important, the reality is that customers doesn’t always send the necessary information on time. If we don’t have the information by the ship by date, we notify the customer and push it out as far as we can.

**Comment by etsyachristensen** [comment]
I believe those values will update. I will have to connect with the shipping/fulfillment team next week to be sure.

**Comment by chris-rogala** [comment]
@etsyachristensen the other thing was the modified date not updating when that ship by date is updated. I can move that to a new issue if that is still not updating.

**Comment by etsyachristensen** [comment]
Yes please do that. It's a separate issue and ultimately that has to go to another team since we aren't in charge of what the internal system does when shipments are updated.

**Comment by chris-rogala** [comment]
@etsyachristensen I will do that and verify the change.

**Comment by glukoid** [comment]
@etsyachristensen

Adding "shipping_method" and "expected_ship_date" to transaction level instead of receipt level means you want to support different shipping methods and date for each transaction in a receipt ?
please correct me if I am wrong here, usually Etsy customer expects to get all ordered items same date.

thx

**Comment by chris-rogala** [comment]
@glukoid I believe the default behavior in Etsy is the latest date wins. Whenever we have a receipt with multiple transactions, the ship by date for the receipt is the latest ship by date. This is why I questioned how it changes when the ship by date is changed on a receipt.

**Comment by glukoid** [comment]
Guys I see request to v3/application/shops/{shopId}/receipts
does not always return correct value for expected_ship_date - time to time it just returned as null from etsy.

bug ? could someone look into it ?

thx

**Comment by chris-rogala** [comment]
@glukoid All of our orders are coming back with `expected_ship_date` being set: 
![image](https://user-images.githubusercontent.com/44238629/134047599-6118af77-b941-4ce9-a2c7-dfc753724dee.png)

I verified all of the current open orders we have.

**Comment by etsyachristensen** [comment]
@glukoid The expected_ship_date will be null if the listing is "dead" or if it has been shipped. As for putting it into the transaction level, this is intentional due to the structure of the system. Each item in an order is assigned a shipping record, regardless of whether it's several items on a receipt or just one. I realize that most of the time, if not all of the time, the values will be the same, but if for some edge case an item gets a different setting on a receipt we don't want one value to override the rest.

@chris-rogala I'm still waiting for an answer on the date getting updated.

**Comment by chris-rogala** [comment]
@etsyachristensen The `shipping_method` is a bit pointless and not really RESTful. The issue is that we have some higher dollar listings that include UPS Ground as the standard shipping method while lower price listings have USPS.

![image](https://user-images.githubusercontent.com/44238629/134049086-a2c34605-cee2-4f8c-9a48-cef822137da4.png)


It would be extremely helpful to have something like:

```
{
    "shipping": {
        "shipping_profile_id": 123456
        "upgrade_id": 7890123
        "expected_ship_date": 1633741301
        "actual_ship_date": null
    }
}
```

I'm also very confused why this was added at the transaction level. Currently, when shipping is purchased, all items in the receipt are part of the shipment. Was the thought to add it at the transaction level so items could be shipped separate?

Also, as for the date updating, I confirmed that it did update on an order we changed the date on so no need to look into that further.

**Comment by chris-rogala** [comment]
@etsyachristensen To be clear, when the ship by date is adjusted in Etsy, the `expected_ship_date` does update. However, the `update_timestamp` does not update when this occurs. I'll separate that into another issue. I'll also move my last comment into a new feature request since technically the original request has been completed.

**Comment by glukoid** [comment]
@etsyachristensen & @chris-rogala 

I made same request twice (second one after minute or two) with same input parameters.
first time I got 4 orders some with and some without expected ship date
e.g.:

"shipping_method": "Standard Shipping",
"expected_ship_date": null,
"actual_ship_date": null

for second time request I got same orders but with correct expected ship date.

All those orders are shipped and not cancelled.
